enhancements/operator-conditions.md

```diff
@@ -0,0 +1,188 @@
+
+#### Story 1
+
+As an operator author, I want operators managed by OLM to be able to communicate complex states that OLM cannot infer on its own.
+
+#### Story 2
+
+As an operator author, I want operators managed by OLM to be able to communicate whether or not they are ready to be upgraded.
+
+## Design Details
+
+### OperatorHealth CRD
+
+OLM will create a channel for operators to communicate complex conditions by introducing the `OperatorHealth` [CustomResourceDefinition](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/).
+
+When an operator is installed, OLM will create an `OperatorHealth` CustomResource (CR) for the operator. The operator should only be concerned with updating the status of the `OperatorHealth` that it owns and will have its RBAC scoped so it may only GET, LIST, or UPDATE the `OperatorHealth` CR it is associated with.
+
+The `OperatorHealth` CR will be created with the [aggregation label](https://github.com/openshift/enhancements/blob/master/enhancements/olm/simplify-apis.md#component-selection) introduced for the `OperatorStatus` CRD that will associate it with a specific operator. When OLM deploys the operator, an additional environment variable will be appended to the deployment. Operators will then be able to use this label along with a library provided by the [operator-sdk](https://github.com/operator-framework/operator-sdk) team to set [conditions](https://github.com/kubernetes/enhancements/blob/3aa92e20bdfa2e60e3cb1a2b92cdca61847d7ad2/keps/sig-api-machinery/1623-standardize-conditions/README.md#kep-1623-standardize-conditions) on the CR to communicate complex states to OLM. For those unfamiliar with conditions, they are used by many Kubernetes Resources to provide users with additional insight into the state of the resource and typically adhere to the following format:
+
+| Field | Description |
+| ----- | ----------- |
+| Type | Type of condition in CamelCase or in foo.example.com/CamelCase. |
+| Status | Status of the condition, one of True, False, Unknown. |
+| ObservedGeneration | If set, this represents the .metadata.generation that the condition was set based upon. |
+| LastTransitionTime | Last time the condition transitioned from one status to another. |
+| Reason | The reason for the condition's last transition in CamelCase. |
+| Message | A human readable message indicating details about the transition. |
+
+OLM will define a set of "OLM Supported Conditions" which will influence how OLM manages the Lifecycle of the operator when present.
+
+### OLM Supported Conditions
+
+All "OLM Supported Conditions" will be communicated by setting the conditions in the `status.conditions` array of the `OperatorHealth` CR associated with the operator. The name of each "OLM Supported Condition" will correspond to the `Type` field of the condition added to the array. Let's look at a fake "OLM Supported Condition" named `foo` that has been added to an `OperatorHealth` CR:
+
+```yaml
```

Comment:
should we add a  `spec.type` field to let us discriminate future reporting modes? e.g. `Push` vs. `Pull`

Reply:
At this time it would always default to `Push` correct? Could we add this when we decide to introduce a `Pull` model and make `Push` the default?

Reply:
sure thing.